jfix-zookeeper/src/main/java/ru/fix/zookeeper/lock/PersistentExpiringLockManager.java

```diff
@@ -148,36 +156,171 @@ public void release(LockIdentity lockId) {
+                PersistentExpiringDistributedLock lock,
+                LockProlongationFailedListener listener
+        ) {
+            LockContainer newLockContainer = new LockContainer(lock, listener);
+            LockContainer oldLockContainer = locks.put(lockId, newLockContainer);
+            if (oldLockContainer != null) {
+                return oldLockContainer.lock;
+            } else {
+                return null;
+            }
         }
-    }
 
-    @Override
-    public void close() {
-        locks.forEach((lockId, lockContainer) -> {
-            if (lockId != null && lockContainer != null) {
+        @Nullable
+        public PersistentExpiringDistributedLock removeLock(LockIdentity lockId) {
+            try {
+                globalLock.lock();
+                return Optional.ofNullable(locks.remove(lockId))
+                        .map(lockContainer -> lockContainer.lock)
+                        .orElse(null);
+            } finally {
+                globalLock.unlock();
+            }
+        }
+
+        public boolean contains(LockIdentity lockId) {
+            return locks.containsKey(lockId);
+        }
+
+        public Optional<PersistentExpiringDistributedLock.State> getLockState(LockIdentity lockId) throws Exception {
+            try {
+                globalLock.lock();
+                LockContainer lockContainer = locks.get(lockId);
+                if (lockContainer != null) {
+                    return Optional.of(lockContainer.lock.getState());
+                } else {
+                    return Optional.empty();
+                }
+            } finally {
+                globalLock.unlock();
+            }
+        }
+
+        /**
+         * @param lockProcessor           - processor that returns what to do with Lock in collection
+         * @param lockResultPostProcessor - processor that does some work after processing and applied action
+         */
+        public void processAllLocks(
+                BiFunction<LockIdentity, PersistentExpiringDistributedLock, ProcessingLockResult> lockProcessor,
+                LockResultPostProcessor lockResultPostProcessor
+        ) {
+            locks.forEach((lockId, lockContainer) -> {
                 try {
-                    logger.warn("Active not released lock with lockId={} closed.", lockId);
-                    lockContainer.lock.close();
-                } catch (Exception e) {
-                    logger.error("Failed to close lock with lockId={}", lockId, e);
+                    globalLock.lock();
+                    final ProcessingLockResult processingLockResult;
+                    if (locks.containsKey(lockId)) {
+                        processingLockResult = lockProcessor.apply(lockId, lockContainer.lock);
+                    } else {
+                        processingLockResult = ALREADY_REMOVED;
+                    }
+                    applyProcessingLockResult(lockId, processingLockResult);
```

Comment:
codestyle too, but I think you've extracted the wrong part of the code to function here - the code, changing the class state is hidden somewhere deep, while it is better to see it at once. 
Also, what do this function's parameters mean? "object to apply, operation to do" - it doesn't look good.

Reply:
fixed